dag-cbor-derive/src/gen.rs

```diff
@@ -106,22 +110,38 @@ fn gen_encode_struct_body(s: &Struct) -> TokenStream {
                     None
                 }
             });
-            let fields = s.fields.iter().map(|field| {
-                let key = rename(&field.name, field.rename.as_ref());
-                let binding = &field.binding;
-                default(
-                    binding,
-                    field.default.as_deref(),
-                    quote! {
-                        Encode::encode(#key, c, w)?;
-                        Encode::encode(#binding, c, w)?;
-                    },
-                )
+            let mut cbor_order = s
+                .fields
+                .iter()
+                .map(|field| {
+                    let key = rename(&field.name, field.rename.as_ref());
+                    let binding = &field.binding;
+                    let field = default(
+                        binding,
+                        field.default.as_deref(),
+                        quote! {
+                            Encode::encode(#key, c, w)?;
+                            Encode::encode(#binding, c, w)?;
+                        },
+                    );
+                    (key.to_string(), field)
+                })
+                .collect::<Vec<(String, _)>>();
+            // CBOR RFC-7049 specifies a canonical sort order, where keys are sorted by length
+            // first. This was later revised with RFC-8949, but we need to stick to the original
+            // order to stay compatible with existing data.
+            cbor_order.sort_unstable_by(|(key_a, _), (key_b, _)| {
```

Comment:
I thought about it, but I didn't want to have more complex dependencies/extra crates only because of this short function. So I decided it's small enough to just have it duplicated.